### ISSUE TITLE
Memory Leak using It0, It1, ... Matchers

### ISSUE THREAD

**Issue by Markus-1974** [VSoftTechnologies/Delphi-Mocks#78]
I use the latest Delphi Mocks: 4b52b69, 
additionally Spring4d and DUnitX

In my dpr file, I use 

``` delphi
ReportMemoryLeaksOnShutdown := True;
```

to report memory leaks.

Now, I got the following testing class. 

``` delphi
procedure TestLogger.Setup;
begin
  FTestContainer := TContainer.Create;

  FFilesystemMock := TMock<IFilesystemAdapter>.Create;
  FTestContainer.RegisterType<IFilesystemAdapter>.DelegateTo(
    function: IFilesystemAdapter
    begin
      Result := FFilesystemMock;
    end)
    .AsDefault;

  FTestContainer.RegisterType<TLogger>
    .Implements<ILogger>;

  FTestContainer.Build;


  FLogger := FTestContainer.Resolve<ILogger>;
end;

procedure TestLogger.TearDown;
begin
  FLogger := nil;

  FFilesystemMock.Free;

  FreeAndNil(FTestContainer);
end;

procedure TestLogger.TestLogInfo;
var
  aStream: TTestMemoryStream;
  aFileAsStr: string;
  aTimeStamp: string;
  aDummy: TDateTime;
begin
  // Arrange
  FFilesystemMock.Setup.WillReturn(True).When.DirectoryExists(It0.IsAny<string>, It1.IsAny<Boolean>);

  aStream := TTestMemoryStream.Create;
  aStream.OnDestroy := procedure(ASender: TObject)
    var
      aStringStream: TStringStream;
    begin
      aStringStream := TStringStream.Create;
      try
        aStream.Position := 0;
        aStringStream.CopyFrom(aStream, aStream.Size);
        aFileAsStr := aStringStream.DataString;
      finally
        FreeAndNil(aStringStream);
      end;
    end;

  FFilesystemMock.Setup.WillReturn(TValue.From<TStream>(aStream))
    .When.CreateFileStream(It0.IsAny<string>, It1.IsEqualTo<Word>(fmCreate or fmShareDenyWrite));

  // Act
  FLogger.LogInfo('TEST');

  // Assert
  aTimeStamp := Copy(aFileAsStr, 1, 23);
  Assert.IsTrue(TryStrToDateTime(aTimeStamp, aDummy));
end;
```

Hint: TTestMemoryStream.Create; is not the problem... it is freed by the system under test.

At the end of the testing process, I get the following error:

![leakmsg](https://cloud.githubusercontent.com/assets/15329025/10737688/e17219da-7c13-11e5-8063-55feefc9f27d.JPG)

Is there anything I do wrong at using It0, It1, ... Matchers?
Or does Delphi-Mocks produce those leaks?

**Comment by staticcat** [comment]
Based on what your reporting here it would seem that Delphi-Mocks is producing those leaks. I will have to investigate to see where they are occurring and remove them. 

If anyone does have the time to write up some unit tests or a fix for this please send us the pr. Currently my time is limited and mostly tied up with other projects.

**Comment by fabriciocolombo** [comment]
I think it was solved by issue #75

**Comment by staticcat** [comment]
Thanks for the heads up. I will have to double check and resolve this issue if that is the case. If someone beats me to the punch please let me know.
